UnnecessaryParentheses false positives in 8.42

This must be related to https://github.com/checkstyle/checkstyle/issues/5779 -- this behavior appeared in 8.42:

```bash
$ cat TestClass.java
public class TestClass {
    public static void main(String[] args) {
        boolean a = true;
        int b = 42;
        int c = 42;
        System.out.println(a == (b == c));
    }
}
```

```bash
$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
         <module name="UnnecessaryParentheses" />
    </module>
</module>
```

```bash
$ java -jar checkstyle-8.42-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] /mnt/c/Users/ipono/Downloads/__files/TestClass.java:6:33: Unnecessary parentheses around expression. [UnnecessaryParentheses]
Audit done.
Checkstyle ends with 1 errors.
```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java
@@ -51,11 +51,20 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * something like
  * </p>
  * <pre>
- *     int x = (a + b) + c;</pre>
+ * int x = (a + b) + c; // 1st Case
+ * boolean p = true; // 2nd Case
+ * int q = 4;
+ * int r = 3;
+ * if (p == (q &lt;= r)) {}</pre>
  * <p>
- * In the above case, given that <em>a</em>, <em>b</em>, and <em>c</em> are
+ * In the first case, given that <em>a</em>, <em>b</em>, and <em>c</em> are
  * all {@code int} variables, the parentheses around {@code a + b}
  * are not needed.
+ * In the second case, parentheses are required as <em>q</em>, <em>r</em> are
+ * of type {@code int} and <em>p</em> is of type {@code boolean}
+ * and removing parentheses will give a compile time error. Even if <em>q</em>
+ * and <em>r</em> were {@code boolean} still there will be no violation
+ * raised as check is not "type aware".
  * </p>
  * <ul>
  * <li>
@@ -156,20 +165,21 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * </p>
  * <pre>
  * public int square(int a, int b){
- *   int square = (a * b); //violation
- *   return (square);      //violation
+ *   int square = (a * b); // violation
+ *   return (square);      // violation
  * }
  * int sumOfSquares = 0;
- * for(int i=(0); i&lt;10; i++){          //violation
- *   int x = (i + 1);                  //violation
- *   sumOfSquares += (square(x * x));  //violation
+ * for(int i=(0); i&lt;10; i++){          // violation
+ *   int x = (i + 1);                  // violation
+ *   sumOfSquares += (square(x * x));  // violation
  * }
  * double num = (10.0); //violation
  * List&lt;String&gt; list = Arrays.asList(&quot;a1&quot;, &quot;b1&quot;, &quot;c1&quot;);
  * myList.stream()
- *   .filter((s) -&gt; s.startsWith(&quot;c&quot;)) //violation
+ *   .filter((s) -&gt; s.startsWith(&quot;c&quot;)) // violation
  *   .forEach(System.out::println);
  * int a = 10, b = 12, c = 15;
+ * boolean x = true, y = false, z= true;
  * if ((a &gt;= 0 &amp;&amp; b &lt;= 9)            // violation, unnecessary parenthesis
  *          || (c &gt;= 5 &amp;&amp; b &lt;= 5)    // violation, unnecessary parenthesis
  *          || (c &gt;= 3 &amp;&amp; a &lt;= 7)) { // violation, unnecessary parenthesis
@@ -177,6 +187,12 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * }
  * if ((-a) != -27 // violation, unnecessary parenthesis
  *          &amp;&amp; b &gt; 5) {
+ *     return;
+ * }
+ * if (x==(a &lt;= 15)) { // ok
+ *     return;
+ * }
+ * if (x==(y == z)) { // ok
  *     return;
  * }
  * </pre>
@@ -565,10 +581,19 @@ public class UnnecessaryParenthesesCheck extends AbstractCheck {
     private static boolean checkAroundOperators(DetailAST ast) {
         final int type = ast.getType();
         final DetailAST parent = ast.getParent();
-        return (TokenUtil.isOfType(type, CONDITIONALS_AND_RELATIONAL)
-                    || TokenUtil.isOfType(type, UNARY_AND_POSTFIX))
-                && TokenUtil.isOfType(parent.getType(), CONDITIONALS_AND_RELATIONAL)
-                && isSurrounded(ast);
+        final boolean result;
+        if (TokenUtil.isOfType(type, CONDITIONALS_AND_RELATIONAL)
+                && TokenUtil.isOfType(parent.getType(), TokenTypes.EQUAL, TokenTypes.NOT_EQUAL)
+                && isSurrounded(ast)) {
+            result = false;
+        }
+        else {
+            result = (TokenUtil.isOfType(type, CONDITIONALS_AND_RELATIONAL)
+                        || TokenUtil.isOfType(type, UNARY_AND_POSTFIX))
+                    && TokenUtil.isOfType(parent.getType(), CONDITIONALS_AND_RELATIONAL)
+                    && isSurrounded(ast);
+        }
+        return result;
     }
 
     /**

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/UnnecessaryParenthesesCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/UnnecessaryParenthesesCheck.xml
@@ -26,11 +26,20 @@
  something like
  &lt;/p&gt;
  &lt;pre&gt;
-     int x = (a + b) + c;&lt;/pre&gt;
+ int x = (a + b) + c; // 1st Case
+ boolean p = true; // 2nd Case
+ int q = 4;
+ int r = 3;
+ if (p == (q &amp;lt;= r)) {}&lt;/pre&gt;
  &lt;p&gt;
- In the above case, given that &lt;em&gt;a&lt;/em&gt;, &lt;em&gt;b&lt;/em&gt;, and &lt;em&gt;c&lt;/em&gt; are
+ In the first case, given that &lt;em&gt;a&lt;/em&gt;, &lt;em&gt;b&lt;/em&gt;, and &lt;em&gt;c&lt;/em&gt; are
  all {@code int} variables, the parentheses around {@code a + b}
  are not needed.
+ In the second case, parentheses are required as &lt;em&gt;q&lt;/em&gt;, &lt;em&gt;r&lt;/em&gt; are
+ of type {@code int} and &lt;em&gt;p&lt;/em&gt; is of type {@code boolean}
+ and removing parentheses will give a compile time error. Even if &lt;em&gt;q&lt;/em&gt;
+ and &lt;em&gt;r&lt;/em&gt; were {@code boolean} still there will be no violation
+ raised as check is not "type aware".
  &lt;/p&gt;</description>
          <properties>
             <property default-value="EXPR,IDENT,NUM_DOUBLE,NUM_FLOAT,NUM_INT,NUM_LONG,STRING_LITERAL,LITERAL_NULL,LITERAL_FALSE,LITERAL_TRUE,ASSIGN,BAND_ASSIGN,BOR_ASSIGN,BSR_ASSIGN,BXOR_ASSIGN,DIV_ASSIGN,MINUS_ASSIGN,MOD_ASSIGN,PLUS_ASSIGN,SL_ASSIGN,SR_ASSIGN,STAR_ASSIGN,LAMBDA,TEXT_BLOCK_LITERAL_BEGIN,LAND,LITERAL_INSTANCEOF,GT,LT,GE,LE,EQUAL,NOT_EQUAL,UNARY_MINUS,UNARY_PLUS,INC,DEC,LNOT,BNOT,POST_INC,POST_DEC"

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheckTest.java
@@ -215,6 +215,12 @@ public class UnnecessaryParenthesesCheckTest extends AbstractModuleTestSupport {
             "98:13: " + getCheckMessage(MSG_EXPR),
             "99:21: " + getCheckMessage(MSG_EXPR),
             "102:13: " + getCheckMessage(MSG_EXPR),
+            "114:21: " + getCheckMessage(MSG_EXPR),
+            "120:13: " + getCheckMessage(MSG_EXPR),
+            "121:30: " + getCheckMessage(MSG_EXPR),
+            "122:28: " + getCheckMessage(MSG_EXPR),
+            "124:21: " + getCheckMessage(MSG_EXPR),
+            "135:20: " + getCheckMessage(MSG_EXPR),
         };
 
         verify(checkConfig, getPath("InputUnnecessaryParenthesesIfStatement.java"), expected);

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java
@@ -105,6 +105,48 @@ public class InputUnnecessaryParenthesesIfStatement {
         }
     }
 
+    public void checkBooleanStatements() {
+        boolean a = true;
+        int b = 42;
+        int c = 42;
+        int d = 32;
+        if ((b == c) == a
+                && (( // violation, unnecessary parenthesis, the latter one
+                                (b==c)==(d>=b)==a!=(c==d))
+                || (b<=c)!=a==(c>=d))) {
+            return;
+        }
+
+        if (( // violation, unnecessary parenthesis
+                a!=(b==c) && (a // violation, unnecessary parenthesis, before 'a'
+                        && (b==c))) // violation, unnecessary parenthesis, before 'b'
+                 || (a || a!=(b<=c))     // ok
+                 || (a==(b!=d==(c==b) && a!=(b<=c)))) { // violation, unnecessary parenthesis,
+                                                        // after '||'
+            return;
+        }
+
+        if (a==(b>=c && a==(c==d && d!=b)) // ok
+                && a==(c<=d)) { // ok
+           return;
+        }
+
+        if (a && a==(b<=c)==(a
+                && (b<=c))) { // violation, unnecessary parenthesis, before 'b'
+            return;
+        }
+
+        if (a==(b==c) // ok
+                || a!=(b<=c)) { // ok
+            return;
+        }
+
+        if ((b==0) == (c==d) // ok
+                && (Integer.valueOf(d) instanceof Integer) == true) { // ok
+            return;
+        }
+    }
+
 
 }
 

--- a/src/xdocs/config_coding.xml
+++ b/src/xdocs/config_coding.xml
@@ -6857,11 +6857,21 @@ boolean b = (~a) &gt; -27            // parens around ~a
           operator precedence and associativity; therefore it won't catch something like
         </p>
         <source>
-int x = (a + b) + c;
+int x = (a + b) + c; // 1st Case
+boolean p = true; // 2nd Case
+int q = 4;
+int r = 3;
+if (p == (q &lt;= r)) {}
         </source>
         <p>
-          In the above case, given that <em>a</em>, <em>b</em>, and <em>c</em> are all
-          <code>int</code> variables, the parentheses around <code>a + b</code> are not needed.
+          In the first case, given that <em>a</em>, <em>b</em>, and <em>c</em> are all
+          <code>int</code> variables, the parentheses around <code>a + b</code>
+          are not needed.
+          In the second case, parentheses are required as <em>q</em>, <em>r</em> are
+          of type <code>int</code> and <em>p</em> is of type <code>boolean</code>
+          and removing parentheses will give a compile time error. Even if <em>q</em>
+          and <em>r</em> were <code>boolean</code> still there will be no violation
+          raised as check is not "type aware".
         </p>
       </subsection>
 
@@ -7063,20 +7073,21 @@ int x = (a + b) + c;
         </p>
         <source>
 public int square(int a, int b){
-  int square = (a * b); //violation
-  return (square);      //violation
+  int square = (a * b); // violation
+  return (square);      // violation
 }
 int sumOfSquares = 0;
-for(int i=(0); i&lt;10; i++){          //violation
-  int x = (i + 1);                  //violation
-  sumOfSquares += (square(x * x));  //violation
+for(int i=(0); i&lt;10; i++){          // violation
+  int x = (i + 1);                  // violation
+  sumOfSquares += (square(x * x));  // violation
 }
 double num = (10.0); //violation
 List&lt;String&gt; list = Arrays.asList(&quot;a1&quot;, &quot;b1&quot;, &quot;c1&quot;);
 myList.stream()
-  .filter((s) -&gt; s.startsWith(&quot;c&quot;)) //violation
+  .filter((s) -&gt; s.startsWith(&quot;c&quot;)) // violation
   .forEach(System.out::println);
 int a = 10, b = 12, c = 15;
+boolean x = true, y = false, z= true;
 if ((a &gt;= 0 &amp;&amp; b &lt;= 9)            // violation, unnecessary parenthesis
          || (c &gt;= 5 &amp;&amp; b &lt;= 5)    // violation, unnecessary parenthesis
          || (c &gt;= 3 &amp;&amp; a &lt;= 7)) { // violation, unnecessary parenthesis
@@ -7084,6 +7095,12 @@ if ((a &gt;= 0 &amp;&amp; b &lt;= 9)            // violation, unnecessary parent
 }
 if ((-a) != -27 // violation, unnecessary parenthesis
          &amp;&amp; b &gt; 5) {
+    return;
+}
+if (x==(a &lt;= 15)) { // ok
+    return;
+}
+if (x==(y == z)) { // ok
     return;
 }
         </source>

